kubetest: add EKS plugin

gyuho opened the pull request:
https://github.com/aws/awstester implements https://github.com/kubernetes/test-infra/blob/9a88e32c98e4897cdb7ada607d440d2323ad221a/kubetest/main.go#L223-L230

Example Prow configuration for ALB Ingress Controller:

```yaml
presets:
- env:
  - name: AWSTESTER_EKS_KUBETEST_EMBEDDED_BINARY
    value: "true"
  - name: AWSTESTER_EKS_WAIT_BEFORE_DOWN
    value: 1m0s
  - name: AWSTESTER_EKS_DOWN
    value: "true"
  - name: AWSTESTER_EKS_ENABLE_NODE_SSH
    value: "true"
  - name: AWSTESTER_EKS_AWSTESTER_IMAGE
    value: 607362164682.dkr.ecr.us-west-2.amazonaws.com/awstester:180919694a39
  - name: AWSTESTER_EKS_WORKER_NODE_INSTANCE_TYPE
    value: m3.xlarge
  - name: AWSTESTER_EKS_WORKER_NODE_ASG_MIN
    value: "1"
  - name: AWSTESTER_EKS_WORKER_NODE_ASG_MAX
    value: "1"
  - name: AWSTESTER_EKS_ALB_ENABLE
    value: "true"
  - name: AWSTESTER_EKS_ALB_ALB_INGRESS_CONTROLLER_IMAGE
    value: quay.io/coreos/alb-ingress-controller:1.0-beta.7
  - name: AWSTESTER_EKS_ALB_TARGET_TYPE
    value: instance
  - name: AWSTESTER_EKS_ALB_TEST_MODE
    value: nginx
  - name: AWSTESTER_EKS_ALB_TEST_SCALABILITY
    value: "true"
  - name: AWSTESTER_EKS_ALB_TEST_SERVER_REPLICAS
    value: "1"
  - name: AWSTESTER_EKS_ALB_TEST_SERVER_ROUTES
    value: "1"
  - name: AWSTESTER_EKS_ALB_TEST_CLIENTS
    value: "20"
  - name: AWSTESTER_EKS_ALB_TEST_CLIENT_REQUESTS
    value: "2000"
  - name: AWSTESTER_EKS_ALB_TEST_RESPONSE_SIZE
    value: "40960"
  - name: AWSTESTER_EKS_ALB_TEST_CLIENT_ERROR_THRESHOLD
    value: "500"
  - name: AWSTESTER_EKS_ALB_TEST_EXPECT_QPS
    value: "100.000000"
  - name: AWS_SHARED_CREDENTIALS_FILE
    value: /etc/aws-cred-awstester/aws-cred-awstester
  labels:
    preset-e2e-aws-alb-ingress-controller: "true"
  volumeMounts:
  - mountPath: /etc/aws-cred-awstester
    name: aws-cred-awstester
    readOnly: true
  volumes:
  - name: aws-cred-awstester
    secret:
      secretName: aws-cred-awstester

presubmits:
  gyuho/aws-alb-ingress-controller:
  - agent: kubernetes
    always_run: true
    context: test-presubmit
    labels:
      preset-dind-enabled: "true"
      preset-e2e-aws-alb-ingress-controller: "true"
    name: pull-ingress-aws-alb-build
    rerun_command: /test pull-ingress-aws-alb-build
    run_after_success:
    - agent: kubernetes
      always_run: true
      context: test-presubmit
      labels:
        preset-dind-enabled: "true"
        preset-e2e-aws-alb-ingress-controller: "true"
      name: pull-ingress-aws-alb-e2e
      rerun_command: /test pull-ingress-aws-alb-e2e
      spec:
        containers:
        - command:
          - tests/alb-e2e.sh
          image: 607362164682.dkr.ecr.us-west-2.amazonaws.com/awstester:db7d4d8ed97c
          name: ""
          resources: {}
          securityContext:
            privileged: true
      trigger: (?m)^/test pull-ingress-aws-alb-e2e
    spec:
      containers:
      - command:
        - /bin/echo
        - pull-ingress-aws-alb-build
        image: alpine
        name: ""
        resources: {}
        securityContext:
          privileged: true
    trigger: (?m)^/test pull-ingress-aws-alb-build
```

The configuration above has been working well with our internal Prow cluster (see https://github.com/gyuho/aws-alb-ingress-controller/pull/3 for example output).

Now, we would like to integrate with upstream.

Action items (with following PRs):
- Change `DumpClusterLogs` and `Publish` methods to use upstream Google Cloud Storage?
- Mount AWS test account credential to upstream Prow cluster
- Figure out how to use "dind" for building [aws-alb-ingress-controller](https://github.com/kubernetes-sigs/aws-alb-ingress-controller) images
  - Just like [containerd-cri-presubmit-jobs.yaml](https://github.com/kubernetes/test-infra/blob/master/config/jobs/containerd/cri/containerd-cri-presubmit-jobs.yaml)
- Add Prow job for [aws-alb-ingress-controller](https://github.com/kubernetes-sigs/aws-alb-ingress-controller)
- Add [aws-alb-ingress-controller](https://github.com/kubernetes-sigs/aws-alb-ingress-controller) tests to upstream [test grid](https://github.com/kubernetes/test-infra/tree/master/testgrid)
- Add Prow job for [aws-ebs-csi-driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver)
- Add [aws-ebs-csi-driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver) tests to upstream [test grid](https://github.com/kubernetes/test-infra/tree/master/testgrid)
- Add Prow job for [cloud-provider-aws](https://github.com/kubernetes/cloud-provider-aws)
- Add [cloud-provider-aws](https://github.com/kubernetes/cloud-provider-aws) tests to upstream [test grid](https://github.com/kubernetes/test-infra/tree/master/testgrid)
- Add `pull-kubernetes-e2e-aws-eks` Prow job

krzyzacy commented:
/assign
remove WIP when you think this is ready for review

gyuho commented:
@krzyzacy Please review :) And let me know if you have any feedback!

krzyzacy commented on kubetest/eks/eks.go:
```diff
@@ -0,0 +1,237 @@
+}
+
+var (
+	timeout = 3 * time.Hour
```
kubetest has a timeout flag, just pass it down?

krzyzacy commented on kubetest/eks/eks.go:
```diff
@@ -0,0 +1,237 @@
+
+var (
+	timeout = 3 * time.Hour
+	verbose = true
```
(also verbose flag)

krzyzacy commented on kubetest/eks/eks.go:
```diff
@@ -0,0 +1,237 @@
+	return tr.cfg.ClusterState.Created, nil
+}
+
+// DumpClusterLogs uploads local cluster logs to S3.
```
humm our infra already supports upload logs to gcs, which works with gubernator & testgrid etc.. I would dump the log to local $ARTIFACTS and let our tooling layer handle that, otherwise you can also choose to support make test-infra handle artifacts from s3.

gyuho replied:
We want to use the GCS first. And then migrate later, if needed. Will change the code to dump logs to $ARTIFACTS, and address other comments within this week.

> choose to support make test-infra handle artifacts from s3

Sounds good. We can write some proposal later to do this.

Thanks!

gyuho commented:
@krzyzacy All addressed. PTAL. Thanks!

krzyzacy commented on kubetest/eks/eks.go:
```diff
@@ -0,0 +1,272 @@
+func (tr *tester) DumpClusterLogs(localPath, s3Path string) error {
+	if !tr.cfg.KubetestEnableDumpClusterLogs {
+		// let default kubetest log dumper handle all artifact uploads.
+		// See https://github.com/kubernetes/test-infra/pull/9811/files#r225776067.
```
our util will handle your pod logs, but you'll still need to generate things like kubelet logs and master logs somehow :-)

@justinsb already wrote https://github.com/kubernetes/test-infra/blob/master/kubetest/dump.go for kops-aws, maybe you can utilize that?

krzyzacy commented on kubetest/main.go:
```diff
@@ -246,6 +248,20 @@ func getDeployer(o *options) (deployer, error) {
+		cfg.KubetestControlTimeout = timeout
+		cfg.KubetestVerbose = verbose
+		// always let default log dumper handle artifacts
+		cfg.KubetestEnableDumpClusterLogs = false
```
similar to above, I think we should respect the `--dump` flag, and skip dump if that's set to false

krzyzacy commented on kubetest/eks/eks.go:
```diff
@@ -0,0 +1,272 @@
+			time.NewTimer(tr.cfg.KubetestControlTimeout),
+			time.NewTimer(tr.cfg.KubetestControlTimeout),
+			tr.cfg.KubetestVerbose,
+		)
```
why not pass in control from main.go and you can avoid creating new control for all the calls?

krzyzacy commented:
/lint

gyuho commented:
@krzyzacy All addressed. PTAL. Thanks!

krzyzacy commented on kubetest/main.go:
```diff
@@ -246,6 +248,23 @@ func getDeployer(o *options) (deployer, error) {
 	case "gke":
 		return newGKE(o.provider, o.gcpProject, o.gcpZone, o.gcpRegion, o.gcpNetwork, o.gcpNodeImage, o.gcpImageFamily, o.gcpImageProject, o.cluster, &o.testArgs, &o.upgradeArgs)
+	case "eks":
+		cfg := eksconfig.NewDefault()
```
I feel like all the eksconfig call can happen inside the deployer, so main won't have eks dependencies. wdyt

(In the future we will want to provide a deployer register function, and you can move all deployer codes to your k-sigs repo)

gyuho replied:
> you can move all deployer codes to your k-sigs repo

This sounds awesome!

Yes, let me move everything to "kubetest/eks" package, so main.go does not import anything other than  "kubetest/eks".

krzyzacy commented on kubetest/eks/eks.go:
```diff
@@ -0,0 +1,254 @@
+	return err
+}
+
+func (tr *tester) TestALBMetrics() (err error) {
```
also those test functions are not used? plans?

gyuho commented:
PTAL.

@krzyzacy All addressed: only kept the ones needed for deployer interface (`TestALBMetrics` won't be used inside `kubetest` repository so had it defined in the other interface), moved everything into `eks` package, added `OWNER` file with @leakingtapan.

@leakingtapan Added godoc comments to explain `LoadConfig` and `Stop` methods.

Thanks!

leakingtapan commented:
/lgtm

krzyzacy commented:
/lint

krzyzacy commented:
/lgtm
/this-is-unbearable
